sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp

```diff
@@ -1055,6 +1073,39 @@ namespace Azure { namespace Storage { namespace Blobs {
        */
       std::vector<ObjectReplicationRule> Rules;
     };
+    /**
+     * @brief Specifies the immutability policy mode to set on the blob.
+     */
+    class BlobImmutabilityPolicyMode final {
+    public:
+      BlobImmutabilityPolicyMode() = default;
+      explicit BlobImmutabilityPolicyMode(std::string value) : m_value(std::move(value)) {}
```

Comment:
Why do you prefer Foo(std::string f) : m_f(std::move(f) instead of Foo(std::string const& f) : m_f(f) ???

Reply:
I think we've talked about this several times in different occasions.

Basically, since we're defining a constructor here, we don't know what kind of value callers are going to pass, it could be an lvalue or an  rvalue.

In the case of `Foo(std::string f): m_f(std::move(f))`:

a. if parameter is lvalue, we make a copy in `f` then move from `f` to `m_f`.
b. if parameter is rvalue, we move from that rvalue to `f`, then move to `m_f`. No copy is made!

But in the case of `Foo(const std::string& f): m_f(f)`, we make a copy regardless the parameter is lvalue or rvalue.

I'm not the only one that prefers this style. I think Casey is with me. https://github.com/Azure/azure-sdk-for-cpp/pull/2595#discussion_r671440673

There are also a lot of discussion on stackoverflow, like [this](https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference) one

Reply:
I remember, yes. Tks

Reply:
Isn't ServiceVersion exposed to customers?

We expose constants to users to let me choose the API version to send requests to. 

And we show it in the logging policy. 

I would expect customers to ve aware of that.

We promise backwards compatibility to older API in the docs, so we are mentioning there's a service version the SDK is talking to